src/python/dbsClient/apis/dbsClient.py

```diff
@@ -10,6 +10,157 @@
 import sys
 import urllib.request, urllib.parse, urllib.error
 
+def parseStream(results):
+    """"
+    Parse given stream of results
+
+    :param: results: list JSON records
+    :type: list
+    :return: generator of JSON records
+    """
+    for rec in results.split('\n'):
+        if rec:
+            yield json.loads(rec)
+
+def aggAttribute(results, attr):
+    """
+    performs aggregation based on given attribute
+
+    :param: results: list JSON records
+    :type: list
+    :param: attr is name of the attribute in JSON record, e.g. run_num
+    :type: string
+    :return: aggregated list of JSON records
+    """
+    rows = []
+    for row in results:
+        if attr in row and not isinstance(row[attr], list):
+            rows.append(row[attr])
+    if len(rows) > 0:
+        return [{attr: rows}]
+    return results
+
+def aggRuns(results):
+    """
+    performs runs API aggregation
+
+    :param: results: list JSON records
+    :type: list
+    :return: aggregated list of JSON records
+    """
+    return aggAttribute(results, 'run_num')
+
+def aggReleaseVersions(results):
+    """
+    performs release version API aggregation
+
+    :param: results: list JSON records
+    :type: list
+    :return: aggregated list of JSON records
+    """
+    return aggAttribute(results, 'release_version')
+
+def aggDatasetAccessTypes(results):
+    """
+    performs dataset access types API aggregation
+
+    :param: results: list JSON records
+    :type: list
+    :return: aggregated list of JSON records
+    """
+    return aggAttribute(results, 'dataset_access_type')
+
+def aggFileLumis(results):
+    """
+    performs filelumis API aggregation
+
+    :param: results: list JSON records
+    :type: list
+    :return: aggregated list of JSON records
+    """
+    output = []
+    file_run_lumi = {}
+    event_ct = False
+    if results and 'event_count' in results[0]:
+        event_ct = True
+    for rec in results:
+        run = rec['run_num']
+        lfn = rec['logical_file_name']
+        lumi = rec['lumi_section_num']
+        if isinstance(lumi, list):
+            break
+        if event_ct:
+            file_run_lumi.setdefault((lfn, run), []).append([rec['lumi_section_num'], rec['event_count']])
+        else:
+            file_run_lumi.setdefault((lfn, run), []).append(rec['lumi_section_num'])
+    # if we get results from Python server no aggregation is required
```

Comment:
To make it deliver a better performance while we are on the Python based service, how about we move this check to the beginning of this function?

Reply:
no, I can't move it above since I need to get aggregated output (`file_run_lumi`) for `dbs2go` source. This loop will not be execute if data comes from Python server